[2.9.0] Adding docs for agent-tls-mode (#1378)
rancher/rancher-docs: docs/getting-started/installation-and-upgrade/installation-references/tls-settings.md

|-----|-----|-----|-----|
| `CATTLE_TLS_MIN_VERSION` | Minimum TLS version | `1.2` | `1.0`, `1.1`, `1.2`, `1.3` |
| `CATTLE_TLS_CIPHERS` | Allowed TLS cipher suites | `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256`,<br/>`TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384`,<br/>`TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305`,<br/>`TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256`,<br/>`TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384`,<br/>`TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305` | See [Golang tls constants](https://golang.org/pkg/crypto/tls/#pkg-constants) |

## Agent TLS Enforcement

> The `agent-tls-mode` setting controls how Rancher's agents (including cluster-agent, fleet-agent, and system-agent) evaluate Rancher's certificate.

Reviewer: ```suggestion
The `agent-tls-mode` setting controls how Rancher's agents (cluster-agent, fleet-agent, and system-agent) validate Rancher's certificate.
```

> When set to `strict`, Rancher's agents will only trust certificates generated by the Certifcate Authority contained in the `cacerts` setting. 

Reviewer: ```suggestion
When the value is `strict`, Rancher's agents will only trust certificates generated by the Certificate Authority contained in the `cacerts` setting.
```

> When set to `system-store`, Rancher's agents will trust any certificate generated by a public Certificate Authority contained in the OS' trust store (including Authorities such as Let's Encrypt). This can be a security risk, since any certificate generated by these authorities will be viewed as valid.

Reviewer: ```suggestion
When the value is `system-store`, Rancher's agents will trust any certificate generated by a public Certificate Authority contained in the operating system's trust store (including those signed by authorities such as Let's Encrypt). This can be a security risk, since any certificate generated by these authorities will be viewed as valid.
```

> When the value is `system-store`, Rancher's agents will trust any certificate generated by a public Certificate Authority contained in the operating system's trust store (including those signed by authorities such as Let's Encrypt). This can be a security risk, since any certificate generated by these authorities will be viewed as valid.

Author: I've changed "viewed as" to "considered" - the "then" part read a bit awkward to me. If the new version reads poorly to you please re-open the thread and we can discuss.

> While the `strict` option has a higher level of security, it requires Rancher to have access to the CA which generated the certificate visible to the agents. In the case of certain certificate configurations (notably, external certificates), this is not automatic, and extra configuration will be needed. See the [installation guide](../install-upgrade-on-a-kubernetes-cluster/install-upgrade-on-a-kubernetes-cluster.md#3-choose-your-ssl-configuration) for more information on which scenarios require extra configuration.

Reviewer: ```suggestion
While the `strict` option enables a higher level of security, it requires Rancher to have access to the CA which generated the certificate visible to the agents. In the case of certain certificate configurations (notably, external certificates), this is not automatic, and extra configuration will be needed. See the [installation guide](../install-upgrade-on-a-kubernetes-cluster/install-upgrade-on-a-kubernetes-cluster.md#3-choose-your-ssl-configuration) for more information on which scenarios require extra configuration.
```

> This setting defaults to `strict` on new installs of Rancher (starting with v2.9.0), but is set as `system-store` for users installing/upgrading from a prior version.

Reviewer: ```suggestion
This setting defaults to `strict` on new installs of Rancher (starting with v2.9.0), but is set to `system-store` for users installing a prior version or upgrading from it.
```

Author: I made the change on the "to" but I think the original "installing/upgrading from" is more clear than the proposal here.

### Preparing for the Setting Change

> Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If this value is set to true, this indicates the agents for the cluster are ready to operate using the `strict` mode. You can manually inpsect each cluster to see if they are ready using the Rancher UI, or a Kubectl command like the following:

Reviewer: ```suggestion
Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If its value is set to "True", the agents for the cluster are ready to operate using the `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI, or a kubectl command like the following:
```

> Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If its value is set to "True", this indicates the agents for the cluster are ready to operate using the `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI, or a kubectl command like the following:

Author: I took a look here and I don't think that the proposed version reads better - we want the end of the sentence to indicate that there's a specific kubectl command to run (the one below) and it needs to fit with the "using" verb that was provided earlier. I don't think the proposal works with that. I could do something like "by running the following kubectl command", but that is very similar to the initial idea.

```bash
## the below command skips ouputs $CLUSTER_NAME,$STATUS for all non-local clusters
kubectl get cluster.management.cattle.io -o jsonpath='{range .items[?(@.metadata.name!="local")]}{.metadata.name},{.status.conditions[?(@.type=="AgentTlsStrictCheck")].status}{"\n"}{end}'
```

### Changing the Setting

You can change the setting using the Rancher UI or the `agentTLSMode` [helm chart option](./helm-chart-options.md).

:::note

> If the value is set through the helm chart, it must be changed through the helm chart.

Reviewer: ```suggestion
If you specify the value using the Helm chart setting, you may modify it only with Helm.
```

Author: I left off "setting" from the first clause, since I think this is confusing when you bring the v3 setting into the mix.

:::

:::warning

Depending on your cert setup, additional action (such as uploading the Certificate Authority which signed your certs) may be required. Review the [installation guide](../install-upgrade-on-a-kubernetes-cluster/install-upgrade-on-a-kubernetes-cluster.md#3-choose-your-ssl-configuration) before changing the setting to see if any additional requirements apply to your setup.

:::

> To change the option through the UI by navating to the "Global Settings" page, and looking for the `agent-tls-mode` setting near the bottom of the page. When changing the setting through the UI, Rancher will check to ensure that all downstream clusters have the status check set to `true`, in order to avoid outages from a certificate mismatch.

Reviewer: ```suggestion
To change the setting's value through the UI, navigate to the "Global Settings" page and find the `agent-tls-mode` setting near the bottom. When asked to modify the value, Rancher will check that all downstream clusters have the condition `AgentTlsStrictCheck` set to `True`, in order to prevent outages from a certificate mismatch.
```

#### Overriding the Setting Validation Checks

> In some cases, you may want to override the check ensuring all agent's can accept the new TLS configuration. To override this check, you must update the setting using the following steps:

Reviewer: ```suggestion
In some cases, you may want to override the check ensuring all agents can accept the new TLS configuration. To do so, you must update the setting following these steps:
```

> In some cases, you may want to override the check ensuring all agents can accept the new TLS configuration. To do so, you must update the setting using following these steps:

Reviewer 2: ```suggestion
In some cases, you may want to override the check ensuring all agents can accept the new TLS configuration. To do so, you must update the setting using these steps:
```

> In some cases, you may want to override the check ensuring all agents can accept the new TLS configuration. To do so, you must update the setting using following these steps:

Author: I've reworked this sentence a bit to "must update the setting using the following steps" - Like so indicates that there are other alternatives which also work, but that's not true, you need to do the following steps.

:::warning

> Rancher checks the status of all clusters to prevent downstream cluster outages. Overriding this check is not recommended, and should be done with great caution.

Reviewer: ```suggestion
Rancher checks the status of all downstream clusters to prevent outages. Overriding this check is not recommended, and should be done with caution.
```

Author: I left "great" in since I want to emphasize that you really shouldn't be doing this. It's not a "check twice", it's a "don't do this unless you are really sure that you need it".

:::

1. As an admin, generate a kubeconfig for the local cluster. In the below examples, this was saved to the `local_kubeconfig.yaml` file.

> 2. Using the following command, retrieve the current setting, and save it to the `setting.yaml` file:

Reviewer: ```suggestion
2. Using the following command, retrieve the current setting and save it to the `setting.yaml` file:
```

```bash

> kubectl get setting agent-tls-mode -o yaml --kubeconfig=local_kubeconfig.yaml >> setting.yaml

Reviewer 2: Maybe?
```suggestion
kubectl get setting agent-tls-mode -o yaml --kubeconfig=local_kubeconfig.yaml > setting.yaml
```

```

> 3. Update the yaml file, replacing `value` with `strict`. Adding the `cattle.io/force: true` annotation will override the above check, and should only be done with great care:

Reviewer: ```suggestion
3. Update the setting file, replacing `value` with `strict`. Adding the `cattle.io/force: true` annotation will override the cluster condition check, and should only be done with great care:
```

Reviewer: We don't actually check the value of the annotation, only that it's present - the key could have any value.

Author: I think that level of nuance would be too much for these docs (and honestly we should be check for the value set to true).

Author: Right, I think that is something that we should do. But it's separate from this PR. More specifically, the point of this docs is to give people something that they can run to bypass the check. 

What I've provided here does work and is accurate, it's just doesn't explain the full range of options. But I think that's ok - I'm not aware of a reason why a user would want to use anything other than what was provided. At most, I thinnk we would want to include a warning that states that values than `"true"` also work, and you should use caution with this annotation. WDYT?

Reviewer: I just hope people won't try to "turn off" the overriding annotation not by removing it but by setting the value to "false", which will not do what they expect.

Author: Yeah I can add the warning which explains that doesn't work if you would like. That way the solution is clear and precise, but there's a warning about that potential common path.

```yaml
apiVersion: management.cattle.io/v3
customized: false
default: strict
kind: Setting
metadata:
  name: agent-tls-mode
  annotations:
    cattle.io/force: "true"
source: ""

> value: "strict"

Reviewer: ```suggestion
value: strict
```

|-----|-----|-----|-----|
| `CATTLE_TLS_MIN_VERSION` | Minimum TLS version | `1.2` | `1.0`, `1.1`, `1.2`, `1.3` |
| `CATTLE_TLS_CIPHERS` | Allowed TLS cipher suites | `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256`,<br/>`TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384`,<br/>`TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305`,<br/>`TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256`,<br/>`TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384`,<br/>`TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305` | See [Golang tls constants](https://golang.org/pkg/crypto/tls/#pkg-constants) |

## Agent TLS Enforcement

> The `agent-tls-mode` setting controls how Rancher's agents (cluster-agent, fleet-agent, and system-agent) validate Rancher's certificate.

Reviewer 3: ```suggestion
The `agent-tls-mode` setting controls how Rancher's agents (`cluster-agent`, `fleet-agent`, and `system-agent`) validate Rancher's certificate.
```

> When the value is `strict`, Rancher's agents will only trust certificates generated by the Certificate Authority contained in the `cacerts` setting. 

Reviewer 3: ```suggestion
When the value is set to `strict`, Rancher's agents only trust certificates generated by the Certificate Authority contained in the `cacerts` setting. 
```

> When the value is `system-store`, Rancher's agents will trust any certificate generated by a public Certificate Authority contained in the operating system's trust store (including those signed by authorities such as Let's Encrypt). This can be a security risk, since any certificate generated by these authorities will be considered valid.

Reviewer 3: ```suggestion
When the value is set to `system-store`, Rancher's agents trust any certificate generated by a public Certificate Authority contained in the operating system's trust store, including those signed by authorities such as Let's Encrypt. This can be a security risk, since any certificate generated by these authorities are considered valid.
```

Reviewer 4: Perhaps to emphasize the risk?

> since any certificate generated by these external authorities outside the user's control are considered valid.

> When the value is set to `system-store`, Rancher's agents will trust any certificate generated by a public Certificate Authority contained in the operating system's trust store including those signed by authorities such as Let's Encrypt. This can be a security risk, since any certificate generated by these external authorities outside the user's control are considered valid in this state.

Reviewer 3: ```suggestion
When the value is set to `system-store`, Rancher's agents trust any certificate generated by a public Certificate Authority contained in the operating system's trust store, including those signed by authorities such as Let's Encrypt. This can be a security risk, since any certificate generated by these external authorities, which are outside the user's control, are considered valid.
```

> While the `strict` option enables a higher level of security, it requires Rancher to have access to the CA which generated the certificate visible to the agents. In the case of certain certificate configurations (notably, external certificates), this is not automatic, and extra configuration will be needed. See the [installation guide](../install-upgrade-on-a-kubernetes-cluster/install-upgrade-on-a-kubernetes-cluster.md#3-choose-your-ssl-configuration) for more information on which scenarios require extra configuration.

Reviewer 3: ```suggestion
While the `strict` option enables a higher level of security, it requires Rancher to have access to the CA which generated the certificate visible to the agents. In the case of certain certificate configurations (notably, external certificates), this is not automatic, and extra configuration is needed. See the [installation guide](../install-upgrade-on-a-kubernetes-cluster/install-upgrade-on-a-kubernetes-cluster.md#3-choose-your-ssl-configuration) for more information on which scenarios require extra configuration.
```

> This setting defaults to `strict` on new installs of Rancher (starting with v2.9.0), but is set to `system-store` for users installing/upgrading from a prior version.

Reviewer 3: ```suggestion
In Rancher v2.9.0 and later, this setting defaults to `strict` on new installs. It is set to `system-store` for users installing or upgrading from a prior Rancher version.
```

Reviewer 4: What about?

> In Rancher v2.9.0 and later, this setting defaults to `strict` on new installs. For users installing or upgrading from a prior Rancher version, it is set to `system-store`.

### Preparing for the Setting Change

> Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If its value is set to "True", this indicates the agents for the cluster are ready to operate using the `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI or a kubectl command like the following:

Reviewer 3: ```suggestion
Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If `AgentTlsStrictCheck` is set to `True`, this indicates that the agents for the cluster are ready to operate using `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI or a kubectl command such as the following:
```

> Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If `AgentTlsStrictCheck` is set to `"True"`, this indicates that the agents for the cluster are ready to operate in `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI or a kubectl command such as the following:

Reviewer 2: Other places did not have double quotes, maybe consolidate the doc with `True`?

```suggestion
Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If `AgentTlsStrictCheck` is set to `True`, this indicates that the agents for the cluster are ready to operate in `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI or a kubectl command such as the following:
```

Author: I've changed the other line which refers to this condition to use the quotes instead since I think that's more consistent with the state of the code overall. LMK if that doesn't meet with what you were thinking and we can re-evaluate.

> Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If `AgentTlsStrictCheck` is set to `"True"`, this indicates that the agents for the cluster are ready to operate using the `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI or a kubectl command such as the following:

Reviewer 3: ```suggestion
Each cluster contains a condition in the status field called `AgentTlsStrictCheck`. If `AgentTlsStrictCheck` is set to `"True"`, this indicates that the agents for the cluster are ready to operate in `strict` mode. You can manually inspect each cluster to see if they are ready using the Rancher UI or a kubectl command such as the following:
```

```bash
## the below command skips ouputs $CLUSTER_NAME,$STATUS for all non-local clusters
kubectl get cluster.management.cattle.io -o jsonpath='{range .items[?(@.metadata.name!="local")]}{.metadata.name},{.status.conditions[?(@.type=="AgentTlsStrictCheck")].status}{"\n"}{end}'
```

### Changing the Setting

You can change the setting using the Rancher UI or the `agentTLSMode` [helm chart option](./helm-chart-options.md).

:::note

> If you specify the value through the Helm chart, you may only modify the value with helm.

Reviewer 3: ```suggestion
If you specify the value through the Helm chart, you may only modify the value with Helm.
```

:::

:::warning

> Depending on your cert setup, additional action (such as uploading the Certificate Authority which signed your certs) may be required. Review the [installation guide](../install-upgrade-on-a-kubernetes-cluster/install-upgrade-on-a-kubernetes-cluster.md#3-choose-your-ssl-configuration) before changing the setting to see if any additional requirements apply to your setup.

Reviewer 3: ```suggestion
Depending on your cert setup, additional action may be required, such as uploading the Certificate Authority which signed your certs. Review the [installation guide](../install-upgrade-on-a-kubernetes-cluster/install-upgrade-on-a-kubernetes-cluster.md#3-choose-your-ssl-configuration) before changing the setting to see if any additional requirements apply.
```

:::

> To change the setting's value through the UI, navigate to the "Global Settings" page, and find the `agent-tls-mode` setting near the bottom of the page. When you change the setting through the UI, Rancher will first check that all downstream clusters have the condition `AgentTlsStrictCheck` set to `True` before allowing the request, in order to prevent outages from a certificate mismatch.

Reviewer 3: ```suggestion
To change the setting's value through the UI, navigate to the **Global Settings** page, and find the `agent-tls-mode` setting near the bottom of the page. When you change the setting through the UI, Rancher first checks that all downstream clusters have the condition `AgentTlsStrictCheck` set to `True` before allowing the request. This prevents outages from a certificate mismatch.
```

#### Overriding the Setting Validation Checks

> In some cases, you may want to override the check ensuring all agents can accept the new TLS configuration. To do so, you must update the setting using these steps:

Reviewer 3: ```suggestion
In some cases, you may want to override the check ensuring all agents can accept the new TLS configuration:
```

:::warning

Rancher checks the status of all downstream clusters to prevent outages. Overriding this check is not recommended, and should be done with great caution.

:::

1. As an admin, generate a kubeconfig for the local cluster. In the below examples, this was saved to the `local_kubeconfig.yaml` file.

> 2. Using the following command, retrieve the current setting and save it to the `setting.yaml` file:

Reviewer 3: ```suggestion
2. Retrieve the current setting and save it to `setting.yaml`:
```

```bash
kubectl get setting agent-tls-mode -o yaml --kubeconfig=local_kubeconfig.yaml > setting.yaml
```

> 3. Update the `setting.yaml` file, replacing `value` with `strict`. Adding the `cattle.io/force: "true"` annotation will override the cluster condition check, and should only be done with great care:

Reviewer 3: ```suggestion
3. Update the `setting.yaml` file, replacing `value` with `strict`. Adding the `cattle.io/force: "true"` annotation overrides the cluster condition check, and should only be done with great care:
```

Reviewer 4: Is the last `:` instead of `.` right?

Author: Yes, I think in this case it's indicating that the following command is used to perform the action being described in the sentence.

:::warning

> Including the cattle.io/force annotation with any value (including, for example "false") overrides the cluster condition check.

Reviewer 4: ```suggestion
Including the `cattle.io/force` annotation with any value (including, for example `"false"`) overrides the cluster condition check.
```
